File: backend/internal/model/agent.go
package model

import (
	"github.com/yourname/generate-cv/pkg/profileagent"
)

// TokenUsage represents token consumption from Agent API
type TokenUsage = profileagent.TokenUsage

// KeywordReport represents keyword matching analysis
type KeywordReport = profileagent.KeywordReport

// ScoreBreakdown represents per-dimension scores
type ScoreBreakdown = profileagent.ScoreBreakdown

// AgentParseRequest represents a request to parse a CV file
type AgentParseRequest struct {
	Prompt string `form:"prompt" json:"prompt,omitempty"`
	// File is handled via multipart form, not JSON
}

// AgentParseResponse represents the response from parsing a CV file
type AgentParseResponse struct {
	Profile map[string]interface{} `json:"profile"`
	Usage   TokenUsage             `json:"usage"`
}

// AgentEditRequest represents a request to edit a profile
type AgentEditRequest struct {
	Profile     map[string]interface{} `json:"profile" binding:"required"`
	Instruction string                 `json:"instruction" binding:"required"`
}

// AgentEditResponse represents the response from editing a profile
type AgentEditResponse struct {
	Profile map[string]interface{} `json:"profile"`
	Usage   TokenUsage             `json:"usage"`
}

// AgentTailorRequest represents a request to tailor a profile
type AgentTailorRequest struct {
	Profile        map[string]interface{} `json:"profile" binding:"required"`
	JobDescription string                 `json:"job_description" binding:"required"`
	UserPrompt     string                 `json:"user_prompt,omitempty"`
}

// AgentTailorResponse represents the response from tailoring a profile
type AgentTailorResponse struct {
	Profile        map[string]interface{} `json:"profile"`
	KeywordReport  KeywordReport          `json:"keyword_report"`
	RelevanceScore int                    `json:"relevance_score"`
	Usage          TokenUsage             `json:"usage"`
}

// AgentScoreRequest represents a request to score a profile
type AgentScoreRequest struct {
	Profile        map[string]interface{} `json:"profile" binding:"required"`
	JobDescription string                 `json:"job_description,omitempty"`
	UserPrompt     string                 `json:"user_prompt,omitempty"`
}

// AgentScoreResponse represents the response from scoring a profile
type AgentScoreResponse struct {
	OverallScore    int            `json:"overall_score"`
	Breakdown       ScoreBreakdown `json:"breakdown"`
	Recommendations []string       `json:"recommendations"`
	Usage           TokenUsage     `json:"usage"`
}

// AgentPipelineRequest represents a request to run the full pipeline
type AgentPipelineRequest struct {
	JobDescription string `form:"job_description" json:"job_description,omitempty"`
	Prompt         string `form:"prompt" json:"prompt,omitempty"`
	// File is handled via multipart form
}

// AgentPipelineResponse represents the response from running the pipeline
type AgentPipelineResponse struct {
	ParsedProfile   map[string]interface{} `json:"parsed_profile"`
	TailoredProfile map[string]interface{} `json:"tailored_profile,omitempty"`
	KeywordReport   *KeywordReport         `json:"keyword_report,omitempty"`
	RelevanceScore  *int                   `json:"relevance_score,omitempty"`
	Scorecard       *ScoreBreakdown        `json:"scorecard,omitempty"`
	Usage           TokenUsage             `json:"usage"`
}

// ─── AI Inline Suggestions ────────────────────────────────────────────────────

// AIAnalyzeJDRequest is the body for POST /ai/analyze-jd
type AIAnalyzeJDRequest struct {
	CVID           string `json:"cv_id" binding:"required"`
	JobDescription string `json:"job_description" binding:"required"`
}

// AIAnalyzeJDResponse is returned by POST /ai/analyze-jd
type AIAnalyzeJDResponse struct {
	Keywords        []string `json:"keywords"`
	MissingKeywords []string `json:"missing_keywords"`
	MatchScore      int      `json:"match_score"`
	Suggestions     []string `json:"suggestions"`
}

// AISuggestSummaryRequest is the body for POST /ai/suggest-summary
type AISuggestSummaryRequest struct {
	CVID            string `json:"cv_id" binding:"required"`
	JobTitle        string `json:"job_title,omitempty"`
	YearsExperience int    `json:"years_experience,omitempty"`
}

// AISuggestExperienceRequest is the body for POST /ai/suggest-experience
type AISuggestExperienceRequest struct {
	CVID               string `json:"cv_id" binding:"required"`
	Company            string `json:"company" binding:"required"`
	Position           string `json:"position" binding:"required"`
	CurrentDescription string `json:"current_description,omitempty"`
}

// AIRewriteSectionRequest is the body for POST /ai/rewrite-section
type AIRewriteSectionRequest struct {
	CVID      string `json:"cv_id" binding:"required"`
	SectionID string `json:"section_id" binding:"required"`
	Content   string `json:"content" binding:"required"`
	Tone      string `json:"tone" binding:"required"` // professional | concise | impactful
}

// AISuggestionResponse is the generic single-text response for AI suggest endpoints
type AISuggestionResponse struct {
	Suggestion string `json:"suggestion"`
}

File: frontend/lib/api/agent.ts
import apiClient from './client'
import type {
  AgentParseResponse,
  AgentEditRequest,
  AgentEditResponse,
  AgentTailorRequest,
  AgentTailorResponse,
  AgentScoreRequest,
  AgentScoreResponse,
  AgentPipelineResponse,
} from '@/types'

/**
 * Profile Processing Agent API
 * 
 * Endpoints for parsing CVs, editing profiles, tailoring to job descriptions,
 * and scoring candidate fit.
 */
export const agentApi = {
  /**
   * Parse a CV file (PDF, DOCX, MD, TXT) to structured profile
   * @param file - The CV file to parse
   * @param prompt - Optional extraction instructions
   * @returns Parsed profile with token usage
   */
  parse: (file: File, prompt?: string) => {
    const formData = new FormData()
    formData.append('file', file)
    if (prompt) formData.append('prompt', prompt)
    
    return apiClient
      .post<AgentParseResponse>('/agent/parse', formData, {
        headers: { 'Content-Type': 'multipart/form-data' },
        timeout: 60000, // 60s cho parse (file upload + processing)
      })
      .then((r) => r.data)
  },

  /**
   * Edit a profile using natural language instruction
   * @param profile - The structured profile to edit
   * @param instruction - Natural language editing instruction
   * @returns Edited profile with token usage
   */
  edit: (profile: Record<string, unknown>, instruction: string) => {
    const body: AgentEditRequest = { profile, instruction }
    return apiClient
      .post<AgentEditResponse>('/agent/edit', body, { 
        timeout: 30000 // 30s cho edit
      })
      .then((r) => r.data)
  },

  /**
   * Tailor a profile to a job description
   * @param profile - The structured profile to tailor
   * @param jobDescription - The target job description
   * @param userPrompt - Optional additional instructions
   * @returns Tailored profile with keyword report and relevance score
   */
  tailor: (profile: Record<string, unknown>, jobDescription: string, userPrompt?: string) => {
    const body: AgentTailorRequest = { 
      profile, 
      job_description: jobDescription, 
      user_prompt: userPrompt 
    }
    return apiClient
      .post<AgentTailorResponse>('/agent/tailor', body, { 
        timeout: 45000 // 45s cho tailor (phức tạp hơn edit)
      })
      .then((r) => r.data)
  },

  /**
   * Score a profile across 5 dimensions
   * @param profile - The structured profile to score
   * @param jobDescription - Optional job description for context
   * @param userPrompt - Optional scoring focus instructions
   * @returns Score result with breakdown and recommendations
   */
  score: (profile: Record<string, unknown>, jobDescription?: string, userPrompt?: string) => {
    const body: AgentScoreRequest = { 
      profile, 
      job_description: jobDescription, 
      user_prompt: userPrompt 
    }
    return apiClient
      .post<AgentScoreResponse>('/agent/score', body, { 
        timeout: 30000 // 30s cho score
      })
      .then((r) => r.data)
  },

  /**
   * Run full pipeline: parse → tailor (if JD) → score (if JD)
   * @param file - The CV file to process
   * @param jobDescription - Optional job description for tailoring
   * @param prompt - Optional extraction instructions
   * @returns Full pipeline result
   */
  pipeline: (file: File, jobDescription?: string, prompt?: string) => {
    const formData = new FormData()
    formData.append('file', file)
    if (jobDescription) formData.append('job_description', jobDescription)
    if (prompt) formData.append('prompt', prompt)
    
    return apiClient
      .post<AgentPipelineResponse>('/agent/pipeline', formData, {
        headers: { 'Content-Type': 'multipart/form-data' },
        timeout: 120000, // 120s cho pipeline (lâu nhất)
      })
      .then((r) => r.data)
  },

  /**
   * Check Agent service health
   * @returns Health status
   */
  health: () => {
    return apiClient
      .get<{ status: string }>('/agent/health')
      .then((r) => r.data)
  },
}

export default agentApi
